uftrace.c

```diff
@@ -293,19 +293,61 @@ static void parse_debug_domain(char *arg)
 
 static uint64_t parse_time(char *arg)
 {
-	char *unit;
-	uint64_t val = strtoull(arg, &unit, 0);
-
+	char *unit, *pos;
+	int i, multiples = 0;
+	uint64_t val, decimal = 0;
+	char *str = xstrdup(arg);
+
+	pos = strchr(str, '.');
+	if (pos == NULL) {
+		val = strtoull(str, &unit, 0);
+	}
+	else {
+		val = strtoull(str, &pos, 0);
+		decimal = strtoull(++pos, &unit, 0);
+	}
+	/* if the unit is omitted, it is regarded as default unit 'ns'.
+	 * so ignore it below decimal point.
+	 */
 	if (unit == NULL || *unit == '\0')
-		return val;
+		goto out;
 
 	if (!strcasecmp(unit, "us") || !strcasecmp(unit, "usec"))
-		val *= 1000;
+		multiples = 3; /* 1000 */
 	else if (!strcasecmp(unit, "ms") || !strcasecmp(unit, "msec"))
-		val *= 1000 * 1000;
+		multiples = 6; /* 1000 * 1000 */
 	else if (!strcasecmp(unit, "s") || !strcasecmp(unit, "sec"))
-		val *= 1000 * 1000 * 1000;
+		multiples = 9; /* 1000 * 1000 * 1000 */
+	else
+		pr_warn("The unit '%s' isn't supported\n", unit);
+
+	for (i = 0; i < multiples; i++)
+		val *= 10;
+
+	if (decimal) {
+		int cipher = 0, nsec = 0;
+		int decimal_arr[9];
+
+		do {
+			decimal_arr[cipher] = decimal % 10;
+			decimal /= 10;
 
+			if (cipher == 8)
+				break;
+			cipher++;
+		} while (decimal != 0);
+
+		for (i = 0; i < multiples; i++) {
+			nsec = (nsec * 10);
+
+			if (cipher)
+				nsec += decimal_arr[--cipher];
+		}
+
+		val += nsec;
+	}
```

Comment:
Looks complicated.  Also I'd like to share the code with parse_timestamp as much as possible.